formats/img-gif.js

```diff
@@ -0,0 +1,673 @@
+			} else {
+				throw new Error(`This format requires exactly one frame.`);
+			}
+		} else if (!extensionsAllowed && (image.frames.length > 1)) {
+			throw new Error(`This GIF type does not support multiple frames.`);
+		}
+
+		let buffer = new RecordBuffer(image.frames[0].pixels.length);
+
+		for (let i = 0; i < HEADER_MAGIC_LEN; i++) {
+			buffer.write(RecordType.int.u8, headerMagicWord.charCodeAt(i));
+		}
+
+		let globalWidth = (image.width === undefined) ? 0 : image.width;
+		let globalHeight = (image.height === undefined) ? 0 : image.height;
+		for (let f = 0; f < image.frames.length; f++) {
+			if (image.frames[f].width > globalWidth) {
+				globalWidth = image.frames[f].width;
+			}
+			if (image.frames[f].height > globalHeight) {
+				globalHeight = image.frames[f].height;
+			}
+		}
+
+		let logicalDescPackedField = 0x00;
+		let useGlobalColor = false;
+		let globalColorResPower = 0;
+
+		// If we've been provided with a global palette...
+		if (image.palette !== undefined) {
+
+			if (image.palette.length > 256) {
+				throw new Error(`Palette too large; this format supports a maximum of 8bpp of color indexing.`);
+			}
+
+			useGlobalColor = true;
+			logicalDescPackedField |= 0x80;
+			// It seems as though the "Color Resolution" field does not necessarily
+			// have a rigorous definition per the spec, and that it may not be used
+			// by many editors and viewers. It often (but not always?) contains the
+			// same value as the "Size of Global Color Table" field, but the latter
+			// is needed for determining the actual number of bytes occupied by the
+			// the GCT.
+			while ((1 << (globalColorResPower + 1)) < image.palette.length) {
+				globalColorResPower++;
+			}
+			logicalDescPackedField |= (globalColorResPower << 4);
+			logicalDescPackedField |= globalColorResPower;
+		}
+
+		buffer.writeRecord(recordTypes.logicalScreenDescriptor, {
+			canvasWidth: globalWidth,
+			canvasHeight: globalHeight,
+			packedField: logicalDescPackedField,
+			bgColorIndex: 0,
+			pixelAspectRatio: 0,
+		});
+
+		if (useGlobalColor) {
+			for (let p = 0; p < image.palette.length; p++) {
+				buffer.write(RecordType.int.u8, image.palette[p][0]);
+				buffer.write(RecordType.int.u8, image.palette[p][1]);
+				buffer.write(RecordType.int.u8, image.palette[p][2]);
+			}
+		}
+
+		if (extensionsAllowed && (image.animation.length > 0)) {
+			// The only data that we currently support writing to the
+			// Graphic Control Extension is the delay time, and GIF only
+			// supports a single delay time across the entire animation.
+			// Just pick the first delay in the list.
+			buffer.write(RecordType.int.u8, EXTENSION_INTRODUCER);
+			buffer.write(RecordType.int.u8, EXTENSION_TYPE_GFX_CTRL);
+			buffer.writeRecord(recordTypes.gfxCtrlExtension, {
+				byteSize: 0x04,
+				packedField: 0x00,
+				delayTime: Math.floor(image.animation[0].postDelay / 10),
+				transparentColorIndex: 0x00,
```

Comment:
Will this always write palette index 0 as transparent?  Or does it mean 'no transparency'?  I am wondering whether it should scan the incoming image palette, and look for the first entry that is marked as transparent, and use that?  Something like:

    transparentColorIndex: Math.max(image.palette.findIndex(p => p[3] === 0), /* default value: */ 0);

Same for `bgColorIndex` above.  I may be misunderstanding the purpose of these values however.

Also where possible I have been using palette index 255 as the default transparency index since 0 is often used for black.  This is less efficient in some formats (like .png where you end up with an array of 256 bytes instead of one byte in the transparency data block) so it's not a rule set in stone, however.

Reply:
The presence or absence of a transparent color in the Global Color Table is actually determined by bit 0 in packedField. I had been leaving transparentColorIndex set to 0x00 just as a default. The latest commit now sets this properly, based on the result of searching palette entries for the first transparent color (as you suggest).